### PR TITLE
Set lower feecap on PoSt messages with low balance

### PR DESCRIPTION
**Pull request by magik6k**
More testing definitely todo

### CODE REVIEW

File: storage/wdpost_run.go [hunk 1]
```diff
@@ -762,14 +763,32 @@ func (s *WindowPoStScheduler) setSender(ctx context.Context, msg *types.Message,
...
 
+	mm := *msg
+
+	mm.GasPremium, err = s.api.GasEstimateGasPremium(ctx, 5, msg.From, msg.GasLimit, types.TipSetKey{})
+	if err != nil {
+		log.Errorf("failed to estimate minimum gas premium: %+v", err)
+		mm.GasPremium = msg.GasPremium
+	}
+
+	mm.GasFeeCap, err = s.api.GasEstimateFeeCap(ctx, &mm, 2, types.EmptyTSK)
```
Comment (Kubuxu) [inline storage/wdpost_run.go:774]: This should for sure be more than 2, if you are swetting Premium for 5 block inclusion.

Comment (magik6k) [reply storage/wdpost_run.go]: Set it a bit higher, bust still pretty low - this is estimating the lowest fee we'll set on the message if we don't have enough funds in any account to set more. If we have enough funds, we'll use the default estimation from GasEstimateMessageGas above

File: storage/addresses.go [hunk 2]
```diff
@@ -28,20 +28,23 @@ type addrSelectApi interface {
...
+func AddressFor(ctx context.Context, a addrSelectApi, mi miner.MinerInfo, use AddrUse, goodFunds, minFunds abi.TokenAmount) (address.Address, abi.TokenAmount, error) {
 	switch use {
 	case PreCommitAddr, CommitAddr:
 		// always use worker, at least for now
-		return mi.Worker, nil
+		return mi.Worker, big.Zero(), nil
 	}
 
+	leastBad := address.Undef
+	bestAvail := minFunds
```
Comment (whyrusleeping) [inline storage/addresses.go:39]: I feel like this method could be written more cleanly as something like:

```go
for _,a := range append(controlAddrs, owner, worker) {
  ok, err := maybeUseAddress(a, &leastBad, &bestAvail)
  if ok {
   return leastBad, bestAvail
  }
}
```

**Review by arajasek — COMMENTED** [review]
This should work, but I'm confused about the logic. We seem to be doing:

```
- if we find any address with more than `goodFunds`, use that address
- if not, of all the addresses that have more than `minFunds`, use the one with most balance
- if no such addresses exist, eff it, use worker (even if there's a control address with greater balance than worker)
```

Why not do:

```
- if we find any address with more than `goodFunds`, use that address
- if not, use the address that has the most balance (regardless of whether it is better than `minFunds` or not)
```

The change would probably be to initialize `bestAvail` to worker's balance at line 39, and remove the worker from the list of addresses we call `maybeUseAddress` with (and remove the special-case in 53-61).

**Review by Kubuxu — APPROVED** [review]

**Review by arajasek — APPROVED** [review]

**Pull request merged**